File: runtime/py/nightseam/runtime/pattern.py
"""The regular subset of ECMAScript Unicode patterns, over scalar code points.

Parsing and matching follow the reference's internal/pattern implementation.
Counted repetition is evaluated as sets of positions, so even enormous nullable
counts converge without expanding the expression or narrowing its domain.
"""

from dataclasses import dataclass
from functools import cache, lru_cache

from .json import dumps

MAX_CODEPOINT = 0x10FFFF
DIGITS = ((48, 57),)
WORDS = ((48, 57), (65, 90), (95, 95), (97, 122))
SPACE = (
    (9, 13),
    (32, 32),
    (160, 160),
    (0x1680, 0x1680),
    (0x2000, 0x200A),
    (0x2028, 0x2029),
    (0x202F, 0x202F),
    (0x205F, 0x205F),
    (0x3000, 0x3000),
    (0xFEFF, 0xFEFF),
)
OUTSIDE = (
    "(?P<",
    "(?P=",
    "(?<=",
    "(?<!",
    "(?<",
    "(?=",
    "(?!",
    "[[:",
    "\\k<",
    "\\p{",
    "\\P{",
    "\\A",
    "\\z",
    "\\Z",
    "\\Q",
    "\\C",
)


def normalized(spans):
    result = []
    for low, high in sorted(spans):
        if result and low <= result[-1][1] + 1:
            result[-1] = (result[-1][0], max(result[-1][1], high))
        else:
            result.append((low, high))
    return tuple(result)


def complement(spans):
    result, start = [], 0
    for low, high in normalized(spans):
        if start < low:
            result.append((start, low - 1))
        start = high + 1
    if start <= MAX_CODEPOINT:
        result.append((start, MAX_CODEPOINT))
    return tuple(result)


@dataclass(frozen=True)
class Node:
    kind: str
    spans: tuple = ()
    children: tuple = ()
    minimum: int = 0
    maximum: int | None = 0
    width: int = 0


def character(spans):
    return Node("c", spans=normalized(spans), width=1)


def join(kind, children):
    if len(children) == 1:
        return children[0]
    widths = [node.width for node in children]
    return Node(kind, children=tuple(children), width=min(widths, default=0) if kind == "|" else sum(widths))


class Parser:
    def __init__(self, source):
        self.source, self.at = source, 0

    def fail(self):
        raise ValueError("pattern " + dumps(self.source) + ": outside Nightseam dialect")

    def peek(self):
        return self.source[self.at : self.at + 1]

    def take(self):
        result = self.peek()
        if not result:
            self.fail()
        self.at += 1
        return result

    def disjunction(self, group=False):
        branches, sequence = [], []
        while self.peek():
            if self.peek() == ")":
                if not group:
                    self.fail()
                self.at += 1
                return join("|", branches + [join("s", sequence)])
            if self.peek() == "|":
                self.at += 1
                branches.append(join("s", sequence))
                sequence = []
                continue
            node, assertion = self.atom()
            quantifier = self.peek()
            if quantifier and quantifier in "*+?{":
                self.at += 1
                if assertion:
                    self.fail()
                if quantifier == "{":
                    minimum = self.decimal()
                    maximum = minimum
                    if self.peek() == ",":
                        self.at += 1
                        maximum = self.decimal() if self.peek() in "0123456789" and self.peek() else None
                    if self.take() != "}" or maximum is not None and minimum > maximum:
                        self.fail()
                else:
                    minimum = int(quantifier == "+")
                    maximum = 1 if quantifier == "?" else None
                if self.peek() == "?":
                    self.at += 1
                node = Node("r", children=(node,), minimum=minimum, maximum=maximum, width=node.width * minimum)
            sequence.append(node)
        if group:
            self.fail()
        return join("|", branches + [join("s", sequence)])

    def decimal(self):
        start = self.at
        while self.peek() and self.peek() in "0123456789":
            self.at += 1
        if start == self.at:
            self.fail()
        # Python integers retain ordering even beyond engine repetition limits.
        value = 0
        for digit in self.source[start : self.at]:
            value = value * 10 + ord(digit) - 48
        return value

    def atom(self):
        if self.source.startswith(OUTSIDE, self.at):
            self.fail()
        char = self.take()
        if char in "^$":
            return Node(char), True
        if char == ".":
            return character(complement(((10, 10), (13, 13), (0x2028, 0x2029)))), False
        if char == "(":
            if self.peek() == "?":
                if not self.source.startswith("?:", self.at):
                    self.fail()
                self.at += 2
            return self.disjunction(True), False
        if char == "[":
            return character(self.character_class()), False
        if char == "\\":
            spans, assertion = self.escape(False)
            return (Node(assertion), True) if assertion else (character(spans), False)
        if char in "*+?{}]":
            self.fail()
        return character(((ord(char), ord(char)),)), False

    def character_class(self):
        negated = self.peek() == "^"
        if negated:
            self.at += 1
        spans = []
        while self.peek() and self.peek() != "]":
            left = self.class_atom()
            if self.peek() == "-" and self.at + 1 < len(self.source) and self.source[self.at + 1] != "]":
                self.at += 1
                right = self.class_atom()
                if (
                    len(left) != 1
                    or len(right) != 1
                    or left[0][0] != left[0][1]
                    or right[0][0] != right[0][1]
                    or left[0][0] > right[0][0]
                ):
                    self.fail()
                spans.append((left[0][0], right[0][0]))
            else:
                spans.extend(left)
        if self.take() != "]":
            self.fail()
        return complement(spans) if negated else normalized(spans)

    def class_atom(self):
        char = self.take()
        return self.escape(True)[0] if char == "\\" else ((ord(char), ord(char)),)

    def hex(self, count):
        text = self.source[self.at : self.at + count]
        if len(text) != count or any(char not in "0123456789abcdefABCDEF" for char in text):
            self.fail()
        self.at += count
        return int(text, 16)

    def escape(self, in_class):
        if self.source.startswith(OUTSIDE, self.at - 1):
            self.fail()
        char = self.take()
        classes = {"d": DIGITS, "w": WORDS, "s": SPACE}
        if char.lower() in classes:
            spans = classes[char.lower()]
            return (complement(spans) if char.isupper() else spans), None
        if char in "bB" and not in_class:
            return (), char
        escapes = {"f": 12, "n": 10, "r": 13, "t": 9, "v": 11}
        if char in escapes:
            code = escapes[char]
        elif char == "b" and in_class:
            code = 8
        elif char == "0":
            if self.peek() and self.peek() in "0123456789":
                self.fail()
            code = 0
        elif char == "c":
            letter = self.take()
            if not ("A" <= letter <= "Z" or "a" <= letter <= "z"):
                self.fail()
            code = ord(letter) & 31
        elif char == "x":
            code = self.hex(2)
        elif char == "u":
            if self.peek() == "{":
                self.at += 1
                start = self.at
                while self.peek() and self.peek() in "0123456789abcdefABCDEF":
                    self.at += 1
                text = self.source[start : self.at]
                if not text or self.take() != "}":
                    self.fail()
                code = int(text, 16)
                if code > MAX_CODEPOINT:
                    self.fail()
            else:
                code = self.hex(4)
                if 0xD800 <= code <= 0xDBFF and self.source.startswith("\\u", self.at):
                    end = self.at
                    self.at += 2
                    try:
                        low = self.hex(4)
                    except ValueError:
                        low = -1
                    if 0xDC00 <= low <= 0xDFFF:
                        code = 0x10000 + ((code - 0xD800) << 10) + low - 0xDC00
                    else:
                        self.at = end
        elif char in "^$\\.*+?()[]{}|/" or in_class and char == "-":
            code = ord(char)
        else:
            self.fail()
        return ((code, code),), None


class Pattern:
    def __init__(self, source):
        self.tree = Parser(source).disjunction()

    def matches(self, value):
        def word(at):
            return 0 <= at < len(value) and any(low <= ord(value[at]) <= high for low, high in WORDS)

        def advance(child, positions):
            return frozenset(end for start in positions for end in ends(child, start))

        @cache
        def ends(node, start):
            if node.width > len(value) - start:
                return frozenset()
            kind = node.kind
            if kind == "c":
                return (
                    frozenset((start + 1,))
                    if start < len(value) and any(low <= ord(value[start]) <= high for low, high in node.spans)
                    else frozenset()
                )
            if kind in "^$bB":
                accepted = (
                    start == 0
                    if kind == "^"
                    else start == len(value)
                    if kind == "$"
                    else (word(start - 1) != word(start)) == (kind == "b")
                )
                return frozenset((start,)) if accepted else frozenset()
            if kind == "|":
                return frozenset(end for child in node.children for end in ends(child, start))
            current = frozenset((start,))
            if kind == "s":
                for child in node.children:
                    current = advance(child, current)
                    if not current:
                        break
                return current
            count, child = 0, node.children[0]
            while count < node.minimum and current:
                following = advance(child, current)
                count += 1
                if following == current:
                    count = node.minimum
                    break
                current = following
            result = set(current)
            while current and (node.maximum is None or count < node.maximum):
                following = advance(child, current)
                if following == current:
                    break
                result.update(following)
                current = following
                count += 1
            return frozenset(result)

        return any(ends(self.tree, start) for start in range(len(value) + 1))


@lru_cache(maxsize=256)
def compile_pattern(source):
    return Pattern(source)

File: runtime/cpp/src/pattern.cpp
#include "pattern.hpp"
#include <nightseam/runtime/value.hpp>
#include <algorithm>
#include <cstdint>
#include <limits>
#include <map>
#include <memory>
#include <stdexcept>
#include <utility>
#include <vector>

namespace nightseam::runtime::detail {
namespace {
using Span = std::pair<char32_t, char32_t>;
using Characters = std::vector<Span>;
using Count = std::uint64_t;
constexpr Count limit = std::numeric_limits<Count>::max();

Characters normalized(Characters set) {
    std::sort(set.begin(), set.end());
    Characters result;
    for (auto span : set) {
        if (!result.empty() && span.first <= result.back().second + 1)
            result.back().second = std::max(result.back().second, span.second);
        else result.push_back(span);
    }
    return result;
}
Characters complement(Characters set) {
    Characters result;
    char32_t next = 0;
    for (auto [lo, hi] : normalized(std::move(set))) {
        if (next < lo) result.emplace_back(next, lo - 1);
        next = hi + 1;
    }
    if (next <= 0x10ffff) result.emplace_back(next, 0x10ffff);
    return result;
}
Characters character(char32_t value) { return {{value, value}}; }
const Characters digits{{U'0', U'9'}};
const Characters words{{U'0', U'9'}, {U'A', U'Z'}, {U'_', U'_'}, {U'a', U'z'}};
const Characters whitespace{{0x9, 0xd}, {0x20, 0x20}, {0xa0, 0xa0}, {0x1680, 0x1680},
    {0x2000, 0x200a}, {0x2028, 0x2029}, {0x202f, 0x202f}, {0x205f, 0x205f},
    {0x3000, 0x3000}, {0xfeff, 0xfeff}};

struct Node {
    char kind;
    Characters set;
    std::vector<std::shared_ptr<Node>> children;
    Count min = 0, max = 0, width = 0;
    bool unlimited = false;
    explicit Node(char kind) : kind(kind) {}
};
using Tree = std::shared_ptr<Node>;
Tree set_node(Characters set) {
    auto node = std::make_shared<Node>('c');
    node->set = normalized(std::move(set));
    node->width = 1;
    return node;
}
Tree join(char kind, std::vector<Tree> children) {
    if (children.size() == 1) return children.front();
    auto node = std::make_shared<Node>(kind);
    node->children = std::move(children);
    if (kind == '|') node->width = limit;
    for (const auto& child : node->children) {
        if (kind == '|') node->width = std::min(node->width, child->width);
        else node->width = node->width > limit - child->width ? limit : node->width + child->width;
    }
    return node;
}
Tree repeat(Tree child, Count min, Count max, bool unlimited) {
    auto node = std::make_shared<Node>('r');
    node->width = min != 0 && child->width > limit / min ? limit : child->width * min;
    node->children.push_back(std::move(child));
    node->min = min;
    node->max = max;
    node->unlimited = unlimited;
    return node;
}
[[noreturn]] void invalid() { throw std::invalid_argument("outside Nightseam dialect"); }
bool decimal_digit(char32_t c) { return c >= U'0' && c <= U'9'; }
bool hex_digit(char32_t c) { return decimal_digit(c) || (c >= U'a' && c <= U'f') || (c >= U'A' && c <= U'F'); }
unsigned hex_value(char32_t c) { return decimal_digit(c) ? c - U'0' : (c >= U'a' ? c - U'a' : c - U'A') + 10; }
std::u32string decimal(std::u32string text) {
    auto at = text.find_first_not_of(U'0');
    return at == std::u32string::npos ? U"0" : text.substr(at);
}
Count saturated(const std::u32string& text) {
    Count result = 0;
    for (auto c : text) {
        auto digit = c - U'0';
        if (result > (limit - digit) / 10) return limit;
        result = result * 10 + digit;
    }
    return result;
}

struct Parser {
    std::u32string source;
    std::size_t pos = 0;
    char32_t peek() const { return pos < source.size() ? source[pos] : U'\0'; }
    bool starts(std::u32string_view value) const { return std::u32string_view(source).substr(pos).starts_with(value); }
    void outside() const {
        for (auto prefix : {U"(?P<", U"(?P=", U"(?<=", U"(?<!", U"(?<", U"(?=", U"(?!", U"[[:",
                            U"\\k<", U"\\p{", U"\\P{", U"\\A", U"\\z", U"\\Z", U"\\Q", U"\\C"})
            if (starts(prefix)) invalid();
    }
    std::u32string count() {
        auto start = pos;
        while (pos < source.size() && decimal_digit(source[pos])) ++pos;
        if (pos == start) invalid();
        return decimal(source.substr(start, pos - start));
    }
    Tree disjunction(bool group = false) {
        std::vector<Tree> branches, sequence;
        auto finish = [&] {
            branches.push_back(join('s', std::move(sequence)));
            return join('|', std::move(branches));
        };
        while (pos < source.size()) {
            if (peek() == ')') {
                if (!group) invalid();
                ++pos;
                return finish();
            }
            if (peek() == '|') {
                ++pos;
                branches.push_back(join('s', std::move(sequence)));
                sequence.clear();
                continue;
            }
            auto [node, assertion] = atom();
            if (pos == source.size()) { sequence.push_back(std::move(node)); continue; }
            Count min = 0, max = 1;
            bool unlimited = false;
            if (peek() == '*' || peek() == '+' || peek() == '?') {
                unlimited = peek() != '?';
                min = peek() == '+' ? 1 : 0;
                ++pos;
            } else if (peek() == '{') {
                ++pos;
                auto lower = count();
                auto upper = lower;
                if (peek() == ',') {
                    ++pos;
                    unlimited = true;
                    if (pos < source.size() && decimal_digit(peek())) { upper = count(); unlimited = false; }
                }
                if (peek() != '}' || pos == source.size()) invalid();
                ++pos;
                if (!unlimited && (upper.size() < lower.size() || (upper.size() == lower.size() && upper < lower))) invalid();
                min = saturated(lower);
                max = saturated(upper);
            } else { sequence.push_back(std::move(node)); continue; }
            if (assertion) invalid();
            if (pos < source.size() && peek() == '?') ++pos;
            sequence.push_back(repeat(std::move(node), min, max, unlimited));
        }
        if (group) invalid();
        return finish();
    }
    std::pair<Tree, bool> atom() {
        outside();
        auto c = source[pos++];
        switch (c) {
        case '^': case '$': return {std::make_shared<Node>(static_cast<char>(c)), true};
        case '.': return {set_node(complement({{10, 10}, {13, 13}, {0x2028, 0x2029}})), false};
        case '(':
            if (pos < source.size() && peek() == '?') {
                if (!starts(U"?:")) invalid();
                pos += 2;
            }
            return {disjunction(true), false};
        case '[': return {set_node(character_class()), false};
        case '\\': {
            auto [set, assertion] = escape(false);
            if (assertion != 0) return {std::make_shared<Node>(assertion), true};
            return {set_node(std::move(set)), false};
        }
        case '*': case '+': case '?': case '{': case '}': case ']': invalid();
        default: return {set_node(character(c)), false};
        }
    }
    Characters character_class() {
        bool negated = peek() == '^';
        if (negated) ++pos;
        Characters result;
        while (pos < source.size() && peek() != ']') {
            auto left = class_atom();
            if (pos + 1 < source.size() && peek() == '-' && source[pos + 1] != ']') {
                ++pos;
                auto right = class_atom();
                if (left.size() != 1 || right.size() != 1 || left[0].first != left[0].second ||
                    right[0].first != right[0].second || left[0].first > right[0].first) invalid();
                result.emplace_back(left[0].first, right[0].first);
            } else result.insert(result.end(), left.begin(), left.end());
        }
        if (pos == source.size()) invalid();
        ++pos;
        return negated ? complement(std::move(result)) : result;
    }
    Characters class_atom() {
        auto c = source[pos++];
        return c == '\\' ? escape(true).first : character(c);
    }
    char32_t hex(unsigned count) {
        char32_t value = 0;
        for (unsigned i = 0; i < count; ++i) {
            if (pos == source.size() || !hex_digit(peek())) invalid();
            value = value * 16 + hex_value(source[pos++]);
        }
        return value;
    }
    std::pair<Characters, char> escape(bool in_class) {
        if (pos == source.size()) invalid();
        --pos; outside(); ++pos;
        auto c = source[pos++];
        switch (c) {
        case 'd': return {digits, 0};
        case 'D': return {complement(digits), 0};
        case 'w': return {words, 0};
        case 'W': return {complement(words), 0};
        case 's': return {whitespace, 0};
        case 'S': return {complement(whitespace), 0};
        case 'b': return in_class ? std::pair{character(8), char(0)} : std::pair{Characters{}, 'b'};
        case 'B': if (!in_class) return {{}, 'B'}; break;
        case 'f': return {character(12), 0};
        case 'n': return {character(10), 0};
        case 'r': return {character(13), 0};
        case 't': return {character(9), 0};
        case 'v': return {character(11), 0};
        case '0':
            if (pos < source.size() && decimal_digit(peek())) invalid();
            return {character(0), 0};
        case 'c':
            if (pos < source.size() && ((peek() >= 'A' && peek() <= 'Z') || (peek() >= 'a' && peek() <= 'z')))
                return {character(source[pos++] & 31), 0};
            invalid();
        case 'x': return {character(hex(2)), 0};
        case 'u': {
            if (pos < source.size() && peek() == '{') {
                ++pos;
                auto start = pos;
                char32_t value = 0;
                while (pos < source.size() && hex_digit(peek())) {
                    auto digit = hex_value(source[pos++]);
                    if (value > (0x10ffff - digit) / 16) invalid();
                    value = value * 16 + digit;
                }
                if (pos == start || pos == source.size() || peek() != '}') invalid();
                ++pos;
                return {character(value), 0};
            }
            auto value = hex(4);
            if (value >= 0xd800 && value <= 0xdbff && starts(U"\\u")) {
                auto end = pos;
                pos += 2;
                try {
                    auto trail = hex(4);
                    if (trail >= 0xdc00 && trail <= 0xdfff) value = 0x10000 + (value - 0xd800) * 1024 + trail - 0xdc00;
                    else pos = end;
                } catch (const std::invalid_argument&) { pos = end; }
            }
            return {character(value), 0};
        }
        default:
            if (std::u32string_view(U"^$\\.*+?()[]{}|/").find(c) != std::u32string_view::npos || (in_class && c == '-'))
                return {character(c), 0};
        }
        invalid();
    }
};

using Positions = std::vector<std::size_t>;
void unique(Positions& positions) {
    std::sort(positions.begin(), positions.end());
    positions.erase(std::unique(positions.begin(), positions.end()), positions.end());
}
struct Matcher {
    std::u32string input;
    std::map<std::pair<const Node*, std::size_t>, Positions> memo;
    Positions advance(const Tree& child, const Positions& positions) {
        Positions result;
        for (auto start : positions) {
            auto next = ends(child, start);
            result.insert(result.end(), next.begin(), next.end());
        }
        unique(result);
        return result;
    }
    Positions ends(const Tree& node, std::size_t start) {
        if (node->width > input.size() - start) return {};
        auto key = std::pair{node.get(), start};
        if (auto found = memo.find(key); found != memo.end()) return found->second;
        Positions result;
        switch (node->kind) {
        case 'c':
            if (start < input.size()) {
                for (auto [lo, hi] : node->set) {
                    if (input[start] >= lo && input[start] <= hi) { result.push_back(start + 1); break; }
                }
            }
            break;
        case '^': if (start == 0) result.push_back(start); break;
        case '$': if (start == input.size()) result.push_back(start); break;
        case 'b': case 'B': {
            auto word = [&](std::size_t at) {
                if (at >= input.size()) return false;
                auto c = input[at];
                return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || c == '_' || (c >= 'a' && c <= 'z');
            };
            if (((start > 0 && word(start - 1)) != word(start)) == (node->kind == 'b')) result.push_back(start);
            break;
        }
        case 's':
            result.push_back(start);
            for (const auto& child : node->children) { result = advance(child, result); if (result.empty()) break; }
            break;
        case '|':
            for (const auto& child : node->children) {
                auto next = ends(child, start);
                result.insert(result.end(), next.begin(), next.end());
            }
            unique(result);
            break;
        case 'r': {
            Positions current{start};
            Count count = 0;
            while (count < node->min && !current.empty()) {
                auto next = advance(node->children.front(), current);
                ++count;
                if (next == current) { count = node->min; current = std::move(next); break; }
                current = std::move(next);
            }
            if (!current.empty()) {
                result = current;
                while (node->unlimited || count < node->max) {
                    auto next = advance(node->children.front(), current);
                    if (next.empty() || next == current) break;
                    result.insert(result.end(), next.begin(), next.end());
                    current = std::move(next);
                    ++count;
                }
                unique(result);
            }
            break;
        }
        }
        memo.emplace(key, result);
        return result;
    }
};
}

void check_pattern(std::string_view source) { Parser{decode_utf8(source)}.disjunction(); }
bool match_pattern(std::string_view source, std::string_view value) {
    auto tree = Parser{decode_utf8(source)}.disjunction();
    Matcher matcher{decode_utf8(value), {}};
    for (std::size_t start = 0; start <= matcher.input.size(); ++start)
        if (!matcher.ends(tree, start).empty()) return true;
    return false;
}
} // namespace nightseam::runtime::detail
